### PR TITLE
Add variadic argument type checking

### LINKED ISSUE #1495: Unhandled error with variadic functions
I tried this code:

```rust
fn five() -> i8 {
    5
}

extern "C" {
    fn printf(fmt: *const i8, ...);
}

fn main() {
    let a = 127i8;
    let b = five();

    let c = a + b;

    unsafe { printf("%d\n\0" as *const str as *const i8, c) }
}

```

I expected to see this happen: `rustc` errors out with the following:
```rust
error[E0617]: can't pass `i8` to variadic function
  --> test.rs:15:58
   |
15 |     unsafe { printf("%d\n\0" as *const str as *const i8, c) }
   |                                                          ^ help: cast the value to `c_int`: `c as c_int`
```

Instead, this happened: no error

### PR DESCRIPTION
**Pull request by powerboat9**
Fixes #1495, still need to add tests.

### CODE REVIEW

**Review by CohenArthur — COMMENTED** [review]
Thanks for working on this @powerboat9 :) Where did you find the list of "forbidden" types? Directly in rustc's source code?

One thing I'd like to see is a little bit of documentation explaining the else block. Something simple alongside the lines of "variadic functions cannot take certain types of arguments like i32 and f32" etc. Just to make sure we easily remember what this is for.

We can merge this once you add tests for it.

**Comment by powerboat9** [comment]
Tests added, still unsure whether or not to do ```s/c_int/i32```.

**Comment by CohenArthur** [comment]
@powerboat9 Changing the type to i32, in my opinion, doesn't fix the problem. As mentioned in [the documentation](https://doc.rust-lang.org/core/ffi/type.c_int.html), `c_int` is often `i32` but not always. If you look at the source you'll see the type definition is machine dependent. So hinting to use `i32` instead would cause issues on some weird architectures that `gccrs` will support.

What we should do is use `rust_sorry_at` and mention that we can't yet handle that, and mention that due to some C ABI historical reasons we can't safely pass an actual i16 or i8. And that the user should cast to an integer which is the same size as a C integer on their system.

File: gcc/rust/typecheck/rust-tyty-call.cc [hunk 1]
```diff
@@ -143,6 +143,60 @@ TypeCheckCallExpr::visit (FnType &type)
...
+	    case TyTy::TypeKind::ERROR:
+	      return;
+	      case TyTy::TypeKind::INT: {
+		auto &int_ty
+		  = static_cast<TyTy::IntType &> (*argument_expr_tyty);
+		if ((int_ty.get_int_kind () == TyTy::IntType::IntKind::I8)
+		    || (int_ty.get_int_kind () == TyTy::IntType::IntKind::I16))
+		  {
+		    rust_error_at (arg_locus,
+				   "expected %<c_int%> variadic argument");
```
[the remark this answers is not in the document]
Comment (philberty) [reply gcc/rust/typecheck/rust-tyty-call.cc:159]: awh thats a really good find @powerboat9 leave it as you have it

**Review by philberty — APPROVED** [review]
LGTM

**Pull request merged**